docs/control-structures.rst

```diff
@@ -615,7 +615,13 @@ The following example shows how to use an error string together with ``revert``
         }
     }
 
-The two syntax options are equivalent, it's developer preference which to use.
+If you provide the reason string directly, then the two syntax options are equivalent, it's developer preference which to use.
+
+.. note::
+    The ``require`` function is evaluated just as any other function.
+    This means that all arguments are evaluated before the function itself is executed.
+    In particular, in ``require(condition, f())`` the function ``f`` is executed even if
+    ``condition`` is true.
```

Comment:
I assume that order of argument evaluation is unspecified. However, the following test leads me to believe it is left to right?

```
contract C {
        uint x;
        function setValue() internal returns (string memory s) {
                x = 1337;
        }

        function f() public returns (uint) {
                require(x != 1337 , setValue());
                return x;
        }
}
// ====
// compileViaYul: also
// ----
// f() -> 1337
```

Reply:
It is mostly left to right, but still unspecified.

Reply:
Should we warn for undefined behavior when a value/reference is both read and written across the arguments of require such as in the test case above?

Reply:
If we do that, then we should do it in all functions.

I think we can expect people to be able to grasp the general concept of side-effects and references.

Reply:
Okay, sounds reasonable to me but still somewhat risky. Perhaps we should add such a test case to document current behavior should order of evaluation change at some point in the future?